qemu/tests/numa_cpu.py

```diff
@@ -0,0 +1,149 @@
+            die_min, die_max = _get_boundary(die, vcpu_dies, die_weight)
+            core_min, core_max = _get_boundary(core, vcpu_cores, core_weight)
+            thread_min, thread_max = _get_boundary(thread, vcpu_threads, thread_weight)
+            cpu_min = socket_min + die_min + core_min + thread_min
+            cpu_max = socket_max + die_max + core_max + thread_max
+        elif vm_arch in ('ppc64', 'ppc64le'):
+            cpu_min = int(core)
+            cpu_max = int(core) + vcpu_threads - 1
+        cpu_index = list(range(cpu_min, cpu_max + 1))
+        return cpu_index
+
+    def numa_cpu_guest():
+        """
+        Get the numa cpus for each node in guest os, linux only
+        """
+        error_context.context("Get cpus in guest os", logging.info)
+        numa_info_guest = NumaInfo(session=session)
+        nodes_guest = numa_info_guest.online_nodes
+        numa_cpu_guest = []
+        for node in nodes_guest:
+            numa_cpus = numa_info_guest.online_nodes_cpus[node]
+            numa_cpus = set([int(v) for v in numa_cpus.split()])
+            numa_cpu_guest.append(numa_cpus)
+        return numa_cpu_guest
+
+    def numa_cpu_cli():
+        """
+        Get the numa cpus for each node according to the qemu cli
+        """
+        error_context.context("Get the expected cpus in qemu command line", logging.info)
+        numa_cpus = params.objects("guest_numa_cpus")
+        numa_cpu_cli = []
+        tmp = {}
+        for numa_cpu in numa_cpus:
+            numa_cpu_params = params.object_params(numa_cpu)
+            nodeid = numa_cpu_params["numa_cpu_nodeid"]
+            socket = numa_cpu_params.get("numa_cpu_socketid")
+            die = numa_cpu_params.get("numa_cpu_dieid")
+            core = numa_cpu_params.get("numa_cpu_coreid")
+            thread = numa_cpu_params.get("numa_cpu_threadid")
+            cpu_list = convert_cpu_topology_to_index(socket, die, core, thread)
+            if nodeid in tmp.keys():
+                tmp[nodeid] += cpu_list
+            else:
+                tmp[nodeid] = cpu_list
+        for item in sorted(tmp.items(), key=lambda item: item[0]):
+            numa_cpu_cli.append(set(sorted(item[1])))
+        return numa_cpu_cli
+
+    def get_hotpluggable_cpus():
+        """
+        Parse output of "query-hotpluggable-cpus", get the numa cpus for each node.
+        """
+        error_context.context("Get the hotpluggable cpus", logging.info)
+        specified_cpus = []
+        tmp = {}
+        out = vm.monitor.info("hotpluggable-cpus")
+        for vcpu_info in out:
+            vcpus_count = vcpu_info["vcpus-count"]
+            vcpu_info = vcpu_info["props"]
+            nodeid = vcpu_info.get("node-id")
+            socketid = vcpu_info.get("socket-id")
+            dieid = vcpu_info.get("die-id")
+            coreid = vcpu_info.get("core-id")
+            threadid = vcpu_info.get("thread-id")
```

Comment:
With the tuple I suggested, assuming its name is cpu_topology, L97 ~ L109 can be implemented with below two lines.
```
if nodeid is not None:
  specified_cpus +=  cpu_topology.index([socketid, dieid, coreid, threadid])
```

Reply:
Agree. On x86, can define a CPU topology like:
```
cpu_topo = [[s, d, c, t]
            for s in range(sockets) 
            for d in range(dies) 
            for c in range(cores) 
            for t in range(threads)]
```

Reply:
Thanks all, it is a good method to generate all the cpu topology list, but it does not make this case more easy.
For this case,  we have scenario to test node without thread. Then for multiple threads, it can be [*, *, *, 0], [*, *, *, 1] and so on.

To get the cpu index (cpu id), we need operation to make up the full cpu topology.

Reply:
And, it is different for ppc platform,  to adapt ppc with a this method is also a problem. 
How about keep the current design ? @PaulYuuu  @huangyum

Reply:
For ppc platform, you just need to create a different topology list. But sure, you can keep current design.

Reply:
> For ppc platform, you just need to create a different topology list. But sure, you can keep current design.

Yes, for any design we all can achieve our purpose, the new design seems do not have obvious advantage IMO, let's keep the current design, thanks for your review ~